File: java/stack/lib/src/main/java/org/example/LargestRectangleInHistogram.java
package org.example;

import java.util.ArrayDeque;
import java.util.Deque;

public final class LargestRectangleInHistogram {


    public static int largestRectangleArea(final int[] heights) {
      final Deque<int[]> stack = new ArrayDeque<>(heights.length);
      int maxArea = 0;
      for (int i = 0, start = i; i < heights.length; i++) {
          while (!stack.isEmpty() && stack.peek()[1] > heights[i]) {
              final int[] pop = stack.pop();
              maxArea = Math.max(maxArea, pop[1] * (i - pop[0]));
              start = pop[0];
          }
          stack.push(new int[]{start, heights[i]});
      }
      while (!stack.isEmpty()) {
          final int[] pop = stack.pop();
          maxArea = Math.max(maxArea, pop[1] * (heights.length - pop[0]));
      }
      return maxArea;
    }


}

File: golang/stack/lc_84/largest_rectangle_in_histogram.go
package lc84

func largestRectangleArea(heights []int) int {
	stack, maxArea := [][2]int{}, 0
	for i, v := range heights {
		start := i
		for len(stack) != 0 && stack[len(stack)-1][1] > v {
			pop := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			popArea := pop[1] * (i - pop[0])
			if popArea > maxArea {
				maxArea = popArea
			}
			start = pop[0]
		}
		stack = append(stack, [2]int{start, v})
	}

	for _, pop := range stack {
		popArea := (len(heights) - pop[0]) * pop[1]
		if popArea > maxArea {
			maxArea = popArea
		}
	}
	return maxArea
}
